mindone/models/fit.py

```diff
@@ -0,0 +1,437 @@
+        self.proj = nn.Dense(dim, dim)
+        self.proj_drop = nn.Dropout(p=proj_drop)
+        self.attention = Attention(head_dim, attn_drop=attn_drop)
+
+        self.apply_rotate_embed = apply_rotate_embed
+
+    @staticmethod
+    def _rearange_in(x: Tensor, h: int) -> Tensor:
+        # (b, n, h*d) -> (b, h, n, d)
+        b, n, d = x.shape
+        d = d // h
+
+        x = ops.reshape(x, (b, n, h, d))
+        x = ops.transpose(x, (0, 2, 1, 3))
+        return x
+
+    @staticmethod
+    def _rearange_out(x: Tensor) -> Tensor:
+        # (b, h, n, d) -> (b, n, h*d)
+        b, _, n, _ = x.shape
+        x = ops.transpose(x, (0, 2, 1, 3))
+        x = ops.reshape(x, (b, n, -1))
+        return x
+
+    def construct(self, x: Tensor, mask: Optional[Tensor] = None, freqs_cis: Optional[Tensor] = None) -> Tensor:
+        h = self.num_heads
+        B, N, _ = x.shape
+
+        # (b, n, 3*h*d) -> (b, n, 3, h*d)  -> (3, b, n, h*d)
+        qkv = self.qkv(x).reshape(B, N, 3, -1).permute((2, 0, 1, 3))
+        q, k, v = qkv.unbind(0)
+
+        # (b, n, h*d) -> (b, h, n, d)
+        q = self._rearange_in(q, h)
+        k = self._rearange_in(k, h)
+        v = self._rearange_in(v, h)
+
+        if self.apply_rotate_embed:
+            q, k = apply_rotary_emb(q, k, freqs_cis)
+
+        out = self.attention(q, k, v, mask=mask)
+        # (b, h, n, d) -> (b, n, h*d)
+        out = self._rearange_out(out)
+
+        return self.proj_drop(self.proj(out))
+
+
+class SwiGLU(nn.Cell):
+    def __init__(
+        self,
+        in_features: int,
+        hidden_features: Optional[int] = None,
+        out_features: Optional[int] = None,
+        act_layer: Type[nn.Cell] = nn.SiLU,
+        norm_layer: Optional[Type[nn.Cell]] = None,
+        has_bias: bool = True,
+        drop: float = 0.0,
+    ) -> None:
+        super().__init__()
+        out_features = out_features or in_features
+        hidden_features = hidden_features or in_features
+
+        self.fc1_g = nn.Dense(in_features, hidden_features, has_bias=has_bias)
+        self.fc1_x = nn.Dense(in_features, hidden_features, has_bias=has_bias)
+        self.act = act_layer()
+        self.drop1 = nn.Dropout(p=drop)
+        self.norm = norm_layer((hidden_features,)) if norm_layer is not None else nn.Identity()
+        self.fc2 = nn.Dense(hidden_features, out_features, has_bias=has_bias)
+        self.drop2 = nn.Dropout(p=drop)
+
+    def construct(self, x: Tensor) -> Tensor:
+        x_gate = self.fc1_g(x)
+        x = self.fc1_x(x)
+        x = self.act(x_gate) * x
+        x = self.drop1(x)
+        x = self.norm(x)
+        x = self.fc2(x)
+        x = self.drop2(x)
+        return x
+
+
+class FiTBlock(nn.Cell):
+    """
+    A FiT block with adaptive layer norm zero (adaLN-Zero) conditioning.
+    """
+
+    def __init__(
+        self,
+        hidden_size: int,
+        num_heads: int,
+        mlp_ratio: float = 4.0,
+        ffn: str = "swiglu",
```

Comment:
It will be good to list all possible values:
`ffn: Literal["swiglu", "mlp"] = "swiglu" `

Reply:
python 3.7 does not support literal, but so far mindspore still support 3.7 version..

Reply:
Can do this way:
https://github.com/mindspore-lab/mindone/blob/ead5e008bac80fa5faa9e798ba03d5d5aab268a7/mindone/env/init.py#L5-L8
Sooner or later support for 3.7 will be dropped.

Reply:
great! fixed.